distributed/deploy/subprocess.py

```diff
@@ -15,13 +16,90 @@
 from distributed.compatibility import WINDOWS
 from distributed.deploy.spec import ProcessInterface, SpecCluster
 from distributed.deploy.utils import nprocesses_nthreads
-from distributed.scheduler import Scheduler
 from distributed.worker_memory import parse_memory_limit
 
 logger = logging.getLogger(__name__)
 
 
-class SubprocessWorker(ProcessInterface):
+class Subprocess(ProcessInterface, abc.ABC):
+    process: asyncio.subprocess.Process | None
+
+    def __init__(self):
+        if WINDOWS:
+            # FIXME: distributed#7434
+            raise RuntimeError("Subprocess does not support Windows.")
+        self.process = None
+        super().__init__()
+
+    async def start(self) -> None:
+        await self._start()
+        await super().start()
+
+    @abc.abstractmethod
+    async def _start(self) -> None:
+        """Start the subprocess"""
+
+    async def close(self) -> None:
+        if self.process and self.process.returncode is None:
+            for child in psutil.Process(self.process.pid).children(recursive=True):
+                child.kill()
+            self.process.kill()
+            await self.process.communicate()
+        self.process = None
+        await super().close()
+
+
+class SubprocessScheduler(Subprocess):
+    """A local Dask scheduler running in a dedicated subprocess
+
+    Parameters
+    ----------
+    scheduler_kwargs:
+        Keywords to pass on to the ``Scheduler`` class constructor
+    """
+
+    scheduler_kwargs: dict
+    address: str | None
+
+    def __init__(
+        self,
+        scheduler_kwargs: dict | None = None,
+    ):
+        self.scheduler_kwargs = scheduler_kwargs or {}
+        super().__init__()
+
+    async def _start(self):
+        cmd = [
+            "dask",
+            "spec",
+            "--spec",
+            json.dumps(
+                {"cls": "distributed.Scheduler", "opts": {**self.scheduler_kwargs}}
+            ),
+        ]
+        logger.info(" ".join(cmd))
+        self.process = await asyncio.create_subprocess_exec(
+            "dask",
+            "spec",
+            "--spec",
+            json.dumps(
+                {"cls": "distributed.Scheduler", "opts": {**self.scheduler_kwargs}}
+            ),
+            stderr=asyncio.subprocess.PIPE,
+        )
+
+        while True:
+            line = (await self.process.stderr.readline()).decode()
+            if not line.strip():
+                raise RuntimeError("Scheduler failed to start")
+            logger.info(line.strip())
+            if "Scheduler at" in line:
+                self.address = line.split("Scheduler at:")[1].strip()
+                break
+        logger.debug(line)
```

Comment:
One issue we face here is that the default log level is `WARN`, but we need to enable `INFO` logs on the scheduler to be able to retrieve the scheduler address. I've borrowed this pattern from the SSHCluster implementation, which seemed "good enough". I'm not keen to dive into more work on this for now. , in particular, since it _should_ also entail that we only write logs with the applicable minimum log level.

Reply:
Yeah, that makes sense.